### PR TITLE
Add RFC for workspace mountpoint integration

### PR DESCRIPTION
**Pull request by mmmarcos**

### CODE REVIEW

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+Recently, some changes came to disrupt this stability:
+
+- **Parsec V3 migration**: mountpoint support was completely re-written in Rust
+  which introduced some regressions because:
+  - the new implementation was not identical to the previous one
+  - the tests needed migration too, so edge-cases could have been missed
+  - the 3rd party libraries changed, new issues were detected (particularily on
+    macOS)
+
+- **Mountpoint strategy**: Parsec workspaces can be mounted either as *Drive* or
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:25]: **nitpick (non-blocking):** 
```suggestion
- **Mountpoint strategy**: Parsec workspaces can be mounted either as *Drive* (only on Windows) or
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  as *Directory*. This added some uncertainty as which should be the preferred/
+  default strategy, and if this should be platform-dependent.
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:27]: **note:** Only Windows use a system of drives, so the question is more about the default strategy to use on Windows

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+- **User mount/unmount**: the ability to mount/unmount workspaces was considered
+  not so user-friendly and was removed from the GUI. This introduced some
+  unexpected friction:
+  - a user with too many workspaces is now forced to have all of its workspaces
+    mounted on disk (performance & security concerns arose).
+  - the workspaces "manually" unmounted by the user (via the system's file
+    explorer) are currently not managed by Parsec, leaving space for errors and
+    risk of data loss.
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:36]: **note:** Currently we do not have implemented the `destroy` on `fuser` (currently it's a NOP), there is even a TODO comment:

https://github.com/Scille/parsec-cloud/blob/260577f4da327860b07f6d6ed0935199214c7dbf/libparsec/crates/platform_mountpoint/src/unix/filesystem.rs#L359-L361

Should we create an issue

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+## Goals and Non-Goals
+
+In short, this RFC aims at describing how Parsec should support mountpoint on
+each platform, therefore answering questions such as:
+
+- How workspaces should be mounted by default?
+- What is the expected behavior on startup/shutdown?
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:44]: **nitpick:** 
```suggestion
- What is the expected behavior on startup/shutdown of the application?
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+- How should Parsec recover recover from a mountpoint crash or an application
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:45]: **typo:** 
```suggestion
- How should Parsec recover from a mountpoint crash or an application
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+      (not yet implemented in V3; available in V2 and used by Resana Secure)
+  - :-1: Cons
+    - Needs custom code for integration with File Explorer to support things
+      such as sidebar for quick access, icons, etc.
+    - Directory (either base or workspace dir) can be manually removed as the
+      user sees it as a regular directory
+
+#### Default strategy
+
+On Linux, there is not much difference between Drive and Directory strategies.
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:91]: **note:** The concept of drive (similar to how windows does it) does not exist

Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:91]: **polish:** 
```suggestion
On Unix system (Linux & MacOS), there is not much difference between Drive and Directory strategies.
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+to, need to be mounted according to the mountpoint strategy. Note that the user
+can login into an organization either:
+
+- by selecting an existing organization from the login page
+- by completing the joining process (which triggers a login)
+
+**Workspace access**: If a mountpoint crashes or is manually unmounted by the
+user (via File Explorer) it may not be a good idea to automatically re-mount it
+since this could lead to possibly infinite error loops. Instead, Parsec should
+try to re-mount the workspace whenever the user access the workspace on the GUI.
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:118]: **issue (non-blocking):** IMO that approach is not ideal if the crash is systematic, ideally we would have a retry counter to stop trying after N attempts.

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+Each workspace will be mounted in a `workspace_dir_path` constructed by joining
+`base_mountpoint_dir` and the `workspace_name`.
+
+If the `workspace_dir_path` already exists, it may be "reused" to mount the
+workspace:
+
+1. if the existing directory was manually created by the user
+   - if empty, should be **safe to mount here**
+   - if not empty, **do not mount here** (it may contain user data)
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:132]: **nitpick:** If not empty, it contain user data (we cannot prove that data come from us)

```suggestion
   - if not empty, **do not mount here** (it contain user data)
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+#### Error Management
+
+| Error | Expected behavior |
+| ----- | ----------------- |
+| **Unable to create workspace dir** | If either `base_mountpoint_dir` or `workspace_dir_path` cannot be created (e.g. permissions, disk space, other), the mount should be aborted as there is no way to correctly mount the workspace. |
+| **Workspace unmounted** | If the workspace crashes, or is manually unmounted by the user, no action is performed. The workspace should be re-mounted when the user attemps to access the workspace again from Parsec GUI. |
+| **Workspace directory removed** | If the workspace directory is manually removed by the user, no action is performed. The workspace should be re-mounted when the user attemps to access the workspace again from Parsec GUI. |
+
+In either of the previous cases, Parsec should inform the user abut the issue.
```
Comment (FirelightFlagboy) [inline docs/rfcs/1012-workspace-mountpoint-integration.md:156]: **typo:** 
```suggestion
In either of the previous cases, Parsec should inform the user about the issue.
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md (hunk 8 above)
Comment (mmmarcos) [reply docs/rfcs/1012-workspace-mountpoint-integration.md:132]:
```suggestion
   - if not empty, **do not mount here** (it contains user data)
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md (hunk 2 above)
Comment (mmmarcos) [reply docs/rfcs/1012-workspace-mountpoint-integration.md:27]:
```suggestion
  default strategy to use on Windows.
```

File: docs/rfcs/1012-workspace-mountpoint-integration.md (hunk 3 above)
Comment (mmmarcos) [reply docs/rfcs/1012-workspace-mountpoint-integration.md:36]: OK, please create an issue to track that.

File: docs/rfcs/1012-workspace-mountpoint-integration.md (hunk 7 above)
Comment (mmmarcos) [reply docs/rfcs/1012-workspace-mountpoint-integration.md:118]: Or maybe both? try remount N times + try remount on user access?

@Max-7 any thoughts on this?

File: docs/rfcs/1012-workspace-mountpoint-integration.md (hunk 3 above)
Comment (FirelightFlagboy) [reply docs/rfcs/1012-workspace-mountpoint-integration.md:36]: Opened: https://github.com/Scille/parsec-cloud/issues/9973

**Review by Max-7 — APPROVED** [review]

File: docs/rfcs/1012-workspace-mountpoint-integration.md (hunk 7 above)
Comment (Max-7) [reply docs/rfcs/1012-workspace-mountpoint-integration.md:118]: Could do both. On error : the GUI tries to remount, if it fails it doesn't retry until the user tries to access to workspace. If the user unmounts manually, we mount only on access.
